Python/specialize.c

```diff
@@ -639,16 +646,9 @@ specialize_dict_access(
         cache1->tp_version = type->tp_version_tag;
         cache0->index = (uint16_t)index;
         *instr = _Py_MAKECODEUNIT(values_op, _Py_OPARG(*instr));
-        return 0;
-    }
-    if (type->tp_dictoffset < 0) {
-        SPECIALIZATION_FAIL(base_op, SPEC_FAIL_OUT_OF_RANGE);
-        return 0;
     }
-    if (type->tp_dictoffset > 0) {
-        PyObject **dictptr = (PyObject **) ((char *)owner + type->tp_dictoffset);
-        PyDictObject *dict = (PyDictObject *)*dictptr;
-        if (dict == NULL || !PyDict_CheckExact(dict)) {
+    else {
+        if (!PyDict_CheckExact(dict)) {
```

Comment:
I'm guessing that if we have a class of objects and only *some* of those have a materialized `__dict__` that will wreak havoc with our specialization, right? Because it will flop-flop as it encounters instances with and without `__dict__`. In this case we have to hope that the vast majority don't have a `__dict__`.

Reply:
Yes.
The specialized case is so much faster though, that it might not be any worse than not specializing at all, even in that case.